src/main/scala/taglessfinalpattern/TaglessFinal.scala

```diff
@@ -0,0 +1,195 @@
+package taglessfinalpattern
+
+import cats.{Id, Monad}
+
+import scala.concurrent.Future
+
+/**
+  * Repository and Service implementation using tagless final pattern.
+  * The idea is to make it easier to test our database layer, using Scala’s higher kinded types to abstract
+  * the Future type constructor away from our traits under test.
+  * Intro to tagless final: https://www.basementcrowd.com/2019/01/17/an-introduction-to-tagless-final-in-scala/.
+  * The similar task example https://github.com/LvivScalaClub/cats-playground/blob/master/src/main/scala/BookRepository.scala
+  */
+
+case class User(id: Long, username: String)
+
+case class IotDevice(id: Long, userId: Long, sn: String)
+
+// NOTE: This import bring into the scope implicits that allow you to call .map and .flatMap on the type F[_]
+// and also bring you typeclasses that know how to flatmap (Monad) and map (Functor) over your higher-kinded type.
+import cats.implicits._
+
+trait UserRepository[F[_]] {
+  def registerUser(username: String): F[User]
+
+  def getById(id: Long): F[Option[User]]
+
+  def getByUsername(username: String): F[Option[User]]
+}
+
+class UserRepositoryID extends UserRepository[Id] {
+
+  private var repository: List[User] = List()
+  private var id: Long = 0
+
+  override def registerUser(username: String): Id[User] = repository.find(user => user.username == username) match {
+    case None => {
+      id = id + 1
+      val user = User(id, username)
+      repository = user :: repository
+      user
+    }
+    case Some(user) => user
+  }
+
+  override def getById(id: Long): Id[Option[User]] = repository.find(user => user.id == id) match {
+    case None => None
+    case Some(user) => Some(user)
+  }
```

Comment:
why you need pattern matching here? `repository.find(user => user.id == id)` should be enough